src/graph.rs

```diff
@@ -12,50 +12,116 @@ pub struct Graph<T> {
     edges: HashMap<Rc<T>, HashSet<Rc<T>>>,
 }
 
-impl<T: Eq + PartialEq + Hash> Graph<T> {
+impl<T: Eq + PartialEq + Hash + Clone> Graph<T> {
     pub fn new() -> Graph<T> {
-        todo!();
+        // Should return an empty graph with default values
+        let edges = HashMap::new();
+        Graph { edges }
     }
 
     pub fn vertices(&self) -> Vec<Rc<T>> {
-        todo!();
+        // Iterate through the edges and return the keys in a Vec
+        let mut vertices = Vec::new();
+        for (vertex, _edges) in self.edges.iter() {
+            vertices.push(vertex.clone());
+        }
+
+        vertices
     }
 
     pub fn insert_vertex(&mut self, u: T) {
-        todo!();
+        // Insert a vertex, this will become the root node if no vertex exists
+        if !self.contains_vertex(&u) {
+            self.edges.insert(Rc::new(u), HashSet::<Rc<T>>::new());
+        }
     }
 
-    pub fn insert_edge(&mut self, u: T, v: T) {
-        // node u can already be in the HashMap or it is not in the HashMap
-        todo!();
+    pub fn remove_vertex(&mut self, u: &T) {
+        assert!(self.contains_vertex(u), "Vertex does not exist");
+        self.edges.remove(u);
     }
 
-    pub fn remove_edge(&mut self, u: &T, v: &T) {
-        todo!();
+    pub fn insert_edge(&mut self, u: T, v: T) {
+        if self.contains_vertex(&u) {
+            self.edges.get_mut(&u).unwrap().insert(Rc::new(v));
+        } else {
+            self.insert_vertex(u.clone());
```

Comment:
No need to clone u and v here.
The function signature specifies that it owns u and v.
So after this function both u and v will be dropped.
Rather we can just put them inside the HashMap